lib/Sema/TypeCheckAttr.cpp

```diff
@@ -3578,8 +3587,20 @@ void AttributeChecker::visitDerivativeAttr(DerivativeAttr *attr) {
   auto resultTanType = valueResultConf.getTypeWitnessByName(
       valueResultType, Ctx.Id_TangentVector);
 
+  // Compute the actual differential/pullback type that we use for comparison
+  // with the expected type. We must canonicalize the derivative interface type
+  // before extracting the differential/pullback type from it, so that the
+  // derivative interface type generic signature is available for simplifying
+  // types.
+  CanType canActualResultType = derivativeInterfaceType->getCanonicalType();
```

Comment:
I wonder if it's possible to canonicalize `derivativeInterfaceType` earlier, like:
```diff
-  auto *derivativeInterfaceType = derivative->getInterfaceType()
-      ->castTo<AnyFunctionType>();
+  auto derivativeInterfaceType = dyn_cast<AnyFunctionType>(
+      derivative->getInterfaceType()->getCanonicalType());
```

I tried this on `tensorflow` branch but it didn't work out of the box:
```
/Users/danielzheng/swift-tf/swift/test/AutoDiff/derivative_attr_type_checking.swift:58:25: error: incorrect message found
Assertion failed: ((size_t)sys::locale::columnWidth(I->getText()) == I->getText().size()), function buildFixItLine, file /Users/danielzheng/swift-tf/llvm-project/llvm/lib/Support/SourceMgr.cpp, line 316.
```
I believe this error occurs when diagnostic messages contain archetypes (`τ_0_0`). Filed TF-1073 to track this improvement.